lib/langs.rb

```diff
@@ -28,4 +34,93 @@ class Language
   extend LanguageHelpers
 end
 
+module LanguageVersionHelpers
+
+  private
+
+  def canonical_version(version)
+    version.match(/^\./) ? "0#{version}" : version
+  end
+
+  # Translate version requirement syntax to a String or Array of Strings that Gem::Dependency.new understands
+  # Pipfile and Cargo.toml files support * and ^ (wildcard and caret) requirements, which Ruby does not
+  # See https://www.python.org/dev/peps/pep-0440/#version-matching
+  # And: https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html
+  def translate_requirement(req, language)
+    if language == :python
+      req.sub!('~=', '~>')
+      req.sub!('==', '=')
+    elsif language == :rust
+      req.sub!('~', '~>')
+      req = "^#{req}" unless req =~ /~|<|>|\*|\^|=/ # No comparators means caret requirement
+    end
+
+    begin
+      case req
+      when /\^/
+        # upper bound = take the left most non-zero digit and +1 it
+        req.sub!('^', '')
+        version = Gem::Version.new(req)
+        segments = version.version.split('.')
+        index = segments.find_index {|seg| seg.to_i > 0} # Find the leftmost non-zero digit.
+        index = segments.rindex {|seg| seg.to_i == 0} unless index # If there is none, find the last 0.
+        segments[index] = segments[index].to_i + 1
+        upper_bound = segments[0..index].join('.')
+        [">= #{version.version}", "< #{upper_bound}"]
+      when '*'
+        '>= 0'
+      when /\.\*/
+        if req =~ /!=/
+          # Turn != 1.1.* into >= 1.2, < 1.1
+          req.sub!('.*', '.0')
+          req.sub!('!=', '')
+          v = Gem::Version.new(req) # Create the bumped version using Gem::Version so pre-release handing will work
+          lower_bound = ">= #{v.bump.version}"
+          upper_bound = "< #{v.version}"
+          [lower_bound, upper_bound]
+        else
+          req.sub!('=', '~>') || req = "~> #{req}" # Turn =1.1.* or 1.1.* into ~> 1.1.*
+          req.sub!('.*', '.0') # Turn ~> 1.1.* into ~> 1.1.0
+          req
+        end
+      else
+        req
+      end
+    rescue ArgumentError => e
```

Comment:
In case any of the `Gem::Version.new` trigger an argument error (i.e. when the version requirement is still malformed despite the translations, e.g. when someone has just entered nonsense in their Cargofile/Pipfile).